File: code/jvm/repository-jdbi/src/main/java/pt/isel/repositories/jdbi/users/UserRepositoryJdbi.java
package pt.isel.repositories.jdbi.users;

import org.jdbi.v3.core.Handle;
import pt.isel.domain.security.PasswordValidationInfo;
import pt.isel.domain.users.User;
import pt.isel.repositories.jdbi.utils.JdbiUtils;
import pt.isel.repositories.users.UserRepository;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;
import java.util.Map;

public class UserRepositoryJdbi implements UserRepository {
    private final Handle handle;

    public UserRepositoryJdbi(Handle handle) {
        this.handle = handle;
    }

    @Override
    public User create(String username, PasswordValidationInfo passwordValidationInfo) {
        Long id = JdbiUtils.executeUpdateAndReturnId(handle, """
                INSERT INTO users (username, password_validation)
                VALUES (:username, :password_validation)
                """, JdbiUtils.params(
                "username", username,
                "password_validation", passwordValidationInfo.validationInfo()
        ));
        return new User(id, username, passwordValidationInfo);
    }

    @Override
    public User findById(Long id) {
        return JdbiUtils.executeQueryToSingle(handle, "SELECT * FROM users WHERE id = :id",
                Map.of("id", id), this::mapRowToUser);
    }

    @Override
    public User findByUsername(String username) {
        return JdbiUtils.executeQueryToSingle(handle, "SELECT * FROM users WHERE username = :username",
                Map.of("username", username), this::mapRowToUser);
    }

    @Override
    public boolean hasUsers() {
        return handle.createQuery("SELECT 1 FROM users LIMIT 1")
                .mapTo(Integer.class)
                .findOne()
                .isPresent();
    }

    @Override
    public List<User> findAll() {
        return JdbiUtils.executeQueryToList(handle, "SELECT * FROM users", Map.of(), this::mapRowToUser);
    }

    @Override
    public void save(User entity) {
        JdbiUtils.executeUpdate(handle, """
                UPDATE users
                SET username = :username, password_validation = :password_validation
                WHERE id = :id
                """, JdbiUtils.params(
                "id", entity.id(),
                "username", entity.username(),
                "password_validation", entity.passwordValidation().validationInfo()
        ));
    }

    @Override
    public void deleteById(Long id) {
        JdbiUtils.executeUpdate(handle, "DELETE FROM users WHERE id = :id", Map.of("id", id));
    }

    @Override
    public void clear() {
        JdbiUtils.executeUpdate(handle, "DELETE FROM users", Map.of());
    }

    private User mapRowToUser(ResultSet rs) throws SQLException {
        return new User(
                rs.getLong("id"),
                rs.getString("username"),
                new PasswordValidationInfo(rs.getString("password_validation"))
        );
    }
}

File: code/jvm/repository-jdbi/src/main/kotlin/pt/isel/repositories/jdbi/users/UserRepositoryJdbi.kt
package pt.isel.repositories.jdbi.users

import org.jdbi.v3.core.Handle
import org.slf4j.LoggerFactory
import pt.isel.domain.security.PasswordValidationInfo
import pt.isel.domain.users.User
import pt.isel.repositories.jdbi.utils.executeQueryToList
import pt.isel.repositories.jdbi.utils.executeQueryToSingle
import pt.isel.repositories.jdbi.utils.executeUpdate
import pt.isel.repositories.jdbi.utils.executeUpdateAndReturnId
import pt.isel.repositories.users.UserRepository
import java.sql.ResultSet

class UserRepositoryJdbi(
    private val handle: Handle,
) : UserRepository {
    override fun create(
        username: String,
        passwordValidationInfo: PasswordValidationInfo,
    ): User {
        val id =
            handle.executeUpdateAndReturnId(
                """
                INSERT INTO dbo.users (username, password_validation)
                VALUES (:username, :password_validation)
                """,
                mapOf(
                    "username" to username,
                    "password_validation" to passwordValidationInfo.validationInfo,
                ),
            )
        return User(id, username, passwordValidationInfo)
    }

    override fun findById(id: Long): User? =
        handle.executeQueryToSingle(
            "SELECT * FROM dbo.users WHERE id = :id",
            mapOf("id" to id),
            ::mapRowToUser,
        )

    override fun findByUsername(username: String): User? =
        handle.executeQueryToSingle(
            "SELECT * FROM dbo.users WHERE username = :username",
            mapOf("username" to username),
            ::mapRowToUser,
        )

    override fun hasUsers(): Boolean =
        handle
            .createQuery("SELECT 1 FROM dbo.users LIMIT 1")
            .mapTo(Int::class.java)
            .findOne()
            .isPresent

    override fun findAll(): List<User> =
        handle.executeQueryToList(
            "SELECT * FROM dbo.users",
            mapper = ::mapRowToUser,
        )

    override fun save(entity: User) {
        handle.executeUpdate(
            """
            UPDATE dbo.users
            SET username = :username, password_validation = :password_validation
            WHERE id = :id
            """,
            mapOf(
                "id" to entity.id,
                "username" to entity.username,
                "password_validation" to entity.passwordValidation.validationInfo,
            ),
        )
    }

    override fun deleteById(id: Long) {
        handle.executeUpdate("DELETE FROM dbo.users WHERE id = :id", mapOf("id" to id))
    }

    override fun clear() {
        handle.executeUpdate("DELETE FROM dbo.users")
    }

    private fun mapRowToUser(rs: ResultSet): User =
        User(
            rs.getLong("id"),
            rs.getString("username"),
            PasswordValidationInfo(rs.getString("password_validation")),
        )

    companion object {
        private val logger = LoggerFactory.getLogger(UserRepositoryJdbi::class.java)
    }
}
